pkg/kv/kvserver/batcheval/cmd_delete_range_test.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+// TestDeleteRangeTombstone tests DeleteRange range tombstones directly, using
+// only a Pebble engine.
+//
+// Most MVCC range tombstone logic is tested exhaustively in the MVCC history
+// tests, this just tests the RPC plumbing.
+func TestDeleteRangeTombstone(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	// Initial data for each test: point keys at b@2, c@5, d@2, point tombstone at
+	// d@3, point intent at i@7, and a range tombstone at [k-p)@3.
+	writeInitialData := func(t *testing.T, ctx context.Context, rw storage.ReadWriter) {
+		t.Helper()
+		txn := roachpb.MakeTransaction("test", nil /* baseKey */, roachpb.NormalUserPriority, hlc.Timestamp{WallTime: 7}, 0, 0)
+		require.NoError(t, storage.MVCCPut(ctx, rw, nil, roachpb.Key("b"), hlc.Timestamp{WallTime: 2}, roachpb.MakeValueFromString("b2"), nil))
+		require.NoError(t, storage.MVCCPut(ctx, rw, nil, roachpb.Key("c"), hlc.Timestamp{WallTime: 5}, roachpb.MakeValueFromString("c5"), nil))
+		require.NoError(t, storage.MVCCPut(ctx, rw, nil, roachpb.Key("d"), hlc.Timestamp{WallTime: 2}, roachpb.MakeValueFromString("d2"), nil))
+		require.NoError(t, storage.MVCCDelete(ctx, rw, nil, roachpb.Key("d"), hlc.Timestamp{WallTime: 3}, nil))
+		require.NoError(t, storage.MVCCPut(ctx, rw, nil, roachpb.Key("i"), hlc.Timestamp{WallTime: 7}, roachpb.MakeValueFromString("i7"), &txn))
+		require.NoError(t, storage.ExperimentalMVCCDeleteRangeUsingTombstone(ctx, rw, nil, roachpb.Key("k"), roachpb.Key("p"), hlc.Timestamp{WallTime: 3}, 0))
+	}
+
+	testcases := map[string]struct {
+		start      string
+		end        string
+		ts         int64
+		txn        bool
+		inline     bool
+		returnKeys bool
+		expectErr  interface{} // error type, substring, or true (any)
+	}{
+		"above points succeed": {
+			start:     "a",
+			end:       "f",
+			ts:        10,
+			expectErr: nil,
+		},
+		"above range tombstone succeed": {
+			start:     "k",
+			end:       "z",
+			ts:        10,
+			expectErr: nil,
+		},
+		"transaction errors": {
+			start:     "a",
+			end:       "f",
+			ts:        10,
+			txn:       true,
+			expectErr: batcheval.ErrTransactionUnsupported,
+		},
+		"inline errors": {
+			start:     "a",
+			end:       "f",
+			ts:        10,
+			inline:    true,
+			expectErr: "Inline can't be used with range tombstones",
+		},
+		"returnKeys errors": {
+			start:      "a",
+			end:        "f",
+			ts:         10,
+			returnKeys: true,
+			expectErr:  "ReturnKeys can't be used with range tombstones",
+		},
+		"intent errors with WriteIntentError": {
```

Comment:
Should we test range under intent ts for completeness? Scan intent currently doesn't know about timestamp, but that's an implementation detail so might change in the future.

Reply:
This isn't really meant to be comprehensive, only that errors are plumbed through -- we have comprehensive tests for this in `TestMVCCHistories`, including above/below the intent timestamp (see `range_tombstone_mutations`).